## Lecture 02, Exercise 01 PANDAS Intro

In [1]:
import pandas as pd

In [2]:
students_performance = pd.read_csv("data/StudentsPerformance.csv")

#### Exercise

In the cell below fill in the blank to display general dataframe information.

In [3]:
students_performance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


**Interpretation of data**
 - there are 1000 entries (rows), with default indexing (0-999)
 - All columns have values (non-null)
 - Two main types: 5 object (probably strings) columns and 3 integer (scores) columns
 - Whole dataset is 62.6 kB large

#### Exercise

In the cell below, fill in the blank so that the variable cols stores the df's column names. NOTE: Please keep the type of the data structure as a <class 'pandas.core.indexes.base.Index'>. Do not have to convert this to a list.

In [4]:
cols = students_performance.columns
print(cols)
print("type: ", type(cols))

Index(['gender', 'race/ethnicity', 'parental level of education', 'lunch',
       'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')
type:  <class 'pandas.core.indexes.base.Index'>


#### Exercise

In the cell below, fill in the blank so that:

num_cols stores the number of columns in df

In [5]:
students_performance_shape = students_performance.shape
num_rows = students_performance_shape[0]
num_cols = students_performance_shape[1]

print(f"number of rows: {num_rows}, number of columns: {num_cols}")

number of rows: 1000, number of columns: 8


#### Exercise

In the cell below, fill in the blank so that first_seven is equal to the first 7 rows. 

In [6]:
first_seven = students_performance.head(7)
first_seven

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75  
5                    none          71             83             78  
6               completed          88             95             92

#### Exercise

In the cell below, fill in the blank so that last_four is equal to the last 4 rows. 

In [7]:
last_four = students_performance.tail(4)
last_four

gender race/ethnicity parental level of education         lunch  \
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86

#### Exercise

In the cell below, fill in the blank so that the `unique_parental_education_levels` variable stores a list of the 6 distinct values found within the `parental level of education` column of `df`.

In [8]:
unique_parental_education_levels = students_performance["parental level of education"].unique()
print(unique_parental_education_levels)
print(f"number of parental level of education: {len(unique_parental_education_levels)}")

["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
number of parental level of education: 6


#### Exercise

In the cell below, fill in the blank so that the `scored_100_at_math` variable stores the **DataFrame row(s)** that correspond to everyone who scored 100 at math.

In [9]:
scored_100_at_math = students_performance[students_performance["math score"] == 100]
scored_100_at_math

gender race/ethnicity parental level of education         lunch  \
149    male        group E          associate's degree  free/reduced   
451  female        group E                some college      standard   
458  female        group E           bachelor's degree      standard   
623    male        group A                some college      standard   
625    male        group D                some college      standard   
916    male        group E           bachelor's degree      standard   
962  female        group E          associate's degree      standard   

    test preparation course  math score  reading score  writing score  
149               completed         100            100             93  
451                    none         100             92             97  
458                    none         100            100            100  
623               completed         100             96             86  
625               completed         100             97             99  
916               completed         100            100            100  
962                    none         100            100            100

#### Exercise

In the cell below, fill in the blank to display scores' descriptive statistics

In [10]:
students_performance.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

**Interpretation of data**
 - no missing data
 - Students perform better in reading/writing than in math (greater mean values)

#### Exercise

In the cell below, fill in the blanks so that the `uncompleted_with_good_writing_score` variable stores the **DataFrame rows** that correspond to everyone who hasn't completed the preparation course and there writing score is above the median.

In [11]:
condition_1 = students_performance["test preparation course"] != "completed"
condition_2 = students_performance["writing score"] > students_performance["writing score"].median()
uncompleted_with_good_writing_score = students_performance[condition_1 & condition_2]
uncompleted_with_good_writing_score

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
2    female        group B             master's degree      standard   
4      male        group C                some college      standard   
5    female        group B          associate's degree      standard   
12   female        group B                 high school      standard   
..      ...            ...                         ...           ...   
981    male        group D            some high school      standard   
984  female        group C            some high school      standard   
992  female        group D          associate's degree  free/reduced   
993  female        group D           bachelor's degree  free/reduced   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
2                      none          90             95             93  
4                      none          76             78             75  
5                      none          71             83             78  
12                     none          65             81             73  
..                      ...         ...            ...            ...  
981                    none          81             78             78  
984                    none          74             75             82  
992                    none          55             76             76  
993                    none          62             72             74  
999                    none          77             86             86  

[251 rows x 8 columns]